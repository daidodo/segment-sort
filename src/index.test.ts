import segmentSorter, { Comparator } from './';

const ORIGIN = [
  'D',
  'A',
  '..',
  'BB',
  'c',
  'aA',
  'aa',
  '.',
  'Aa',
  'a',
  'AA',
  'B',
  '~',
  'BBB',
  'Ab',
  '_',
  'd',
  'ab',
  'ä¸­',
  'AB',
  'b',
  'aB',
  'C',
  'dd',
];

function sort(s: string[], comp: Comparator | undefined) {
  return [...s].sort(comp);
}

describe('segmentSorter', () => {
  describe('0', () => {
    const comp = segmentSorter(['_']);
    it('should be undefined', () => {
      expect(comp).toBeUndefined();
    });
    describe('0-1', () => {
      const comp = segmentSorter(['_', 'az']);
      it('should sort strings', () => {
        expect(comp).toBeDefined();
        expect(sort(ORIGIN, comp)).toMatchSnapshot();
      });
      describe('0-1-2', () => {
        const comp = segmentSorter(['_', 'az', 'AZ']);
        it('should sort strings', () => {
          expect(comp).toBeDefined();
          expect(sort(ORIGIN, comp)).toMatchSnapshot();
        });
      });
    });
    describe('0-2', () => {
      const comp = segmentSorter(['_', 'AZ']);
      it('should sort strings', () => {
        expect(comp).toBeDefined();
        expect(sort(ORIGIN, comp)).toMatchSnapshot();
      });
      describe('0-2-1', () => {
        const comp = segmentSorter(['_', 'AZ', 'az']);
        it('should sort strings', () => {
          expect(comp).toBeDefined();
          expect(sort(ORIGIN, comp)).toMatchSnapshot();
        });
      });
    });
    describe('0-12', () => {
      const comp = segmentSorter(['_', 'aA']);
      it('should sort strings', () => {
        expect(comp).toBeDefined();
        expect(sort(ORIGIN, comp)).toMatchSnapshot();
      });
    });
    describe('0-21', () => {
      const comp = segmentSorter(['_', 'Aa']);
      it('should sort strings', () => {
        expect(comp).toBeDefined();
        expect(sort(ORIGIN, comp)).toMatchSnapshot();
      });
    });
  });
  describe('1', () => {
    const comp = segmentSorter(['az']);
    it('should be undefined', () => {
      expect(comp).toBeUndefined();
    });
    describe('1-0', () => {
      const comp = segmentSorter(['az', '_']);
      it('should sort strings', () => {
        expect(comp).toBeDefined();
        expect(sort(ORIGIN, comp)).toMatchSnapshot();
      });
      describe('1-0-2', () => {
        const comp = segmentSorter(['az', '_', 'AZ']);
        it('should sort strings', () => {
          expect(comp).toBeDefined();
          expect(sort(ORIGIN, comp)).toMatchSnapshot();
        });
      });
    });
    describe('1-2', () => {
      const comp = segmentSorter(['az', 'AZ']);
      it('should sort strings', () => {
        expect(comp).toBeDefined();
        expect(sort(ORIGIN, comp)).toMatchSnapshot();
      });
      describe('1-2-0', () => {
        const comp = segmentSorter(['az', 'AZ', '_']);
        it('should sort strings', () => {
          expect(comp).toBeDefined();
          expect(sort(ORIGIN, comp)).toMatchSnapshot();
        });
      });
    });
  });
  describe('2', () => {
    const comp = segmentSorter(['AZ']);
    it('should be undefined', () => {
      expect(comp).toBeUndefined();
    });
    describe('2-0', () => {
      const comp = segmentSorter(['AZ', '_']);
      it('should sort strings', () => {
        expect(comp).toBeDefined();
        expect(sort(ORIGIN, comp)).toMatchSnapshot();
      });
      describe('2-0-1', () => {
        const comp = segmentSorter(['AZ', '_', 'az']);
        it('should sort strings', () => {
          expect(comp).toBeDefined();
          expect(sort(ORIGIN, comp)).toMatchSnapshot();
        });
      });
    });
    describe('2-1', () => {
      const comp = segmentSorter(['AZ', 'az']);
      it('should sort strings', () => {
        expect(comp).toBeDefined();
        expect(sort(ORIGIN, comp)).toMatchSnapshot();
      });
      describe('2-1-0', () => {
        const comp = segmentSorter(['AZ', 'az', '_']);
        it('should sort strings', () => {
          expect(comp).toBeDefined();
          expect(sort(ORIGIN, comp)).toMatchSnapshot();
        });
      });
    });
  });
  describe('12', () => {
    const comp = segmentSorter(['aA']);
    it('should sort strings', () => {
      expect(comp).toBeDefined();
      expect(sort(ORIGIN, comp)).toMatchSnapshot();
    });
    describe('12-0', () => {
      const comp = segmentSorter(['aA', '_']);
      it('should sort strings', () => {
        expect(comp).toBeDefined();
        expect(sort(ORIGIN, comp)).toMatchSnapshot();
      });
    });
  });
  describe('21', () => {
    const comp = segmentSorter(['Aa']);
    it('should sort strings', () => {
      expect(comp).toBeDefined();
      expect(sort(ORIGIN, comp)).toMatchSnapshot();
    });
    describe('21-0', () => {
      const comp = segmentSorter(['Aa', '_']);
      it('should sort strings', () => {
        expect(comp).toBeDefined();
        expect(sort(ORIGIN, comp)).toMatchSnapshot();
      });
    });
  });
});
