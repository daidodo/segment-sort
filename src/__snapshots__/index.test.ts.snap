// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`segmentSorter 0 0-1 0-1-2 should sort strings 1`] = `
[
  ".",
  "..",
  "_",
  "a",
  "aa",
  "ab",
  "aA",
  "aB",
  "b",
  "c",
  "d",
  "dd",
  "A",
  "Aa",
  "Ab",
  "AA",
  "AB",
  "B",
  "BB",
  "BBB",
  "C",
  "D",
  "~",
  "中",
]
`;

exports[`segmentSorter 0 0-1 should sort strings 1`] = `
[
  ".",
  "..",
  "_",
  "a",
  "aa",
  "ab",
  "aA",
  "aB",
  "b",
  "c",
  "d",
  "dd",
  "A",
  "Aa",
  "Ab",
  "AA",
  "AB",
  "B",
  "BB",
  "BBB",
  "C",
  "D",
  "~",
  "中",
]
`;

exports[`segmentSorter 0 0-2 0-2-1 should sort strings 1`] = `
[
  ".",
  "..",
  "_",
  "A",
  "AA",
  "AB",
  "Aa",
  "Ab",
  "B",
  "BB",
  "BBB",
  "C",
  "D",
  "a",
  "aA",
  "aB",
  "aa",
  "ab",
  "b",
  "c",
  "d",
  "dd",
  "~",
  "中",
]
`;

exports[`segmentSorter 0 0-2 should sort strings 1`] = `
[
  ".",
  "..",
  "_",
  "A",
  "AA",
  "AB",
  "Aa",
  "Ab",
  "B",
  "BB",
  "BBB",
  "C",
  "D",
  "a",
  "aA",
  "aB",
  "aa",
  "ab",
  "b",
  "c",
  "d",
  "dd",
  "~",
  "中",
]
`;

exports[`segmentSorter 0 0-12 should sort strings 1`] = `
[
  ".",
  "..",
  "_",
  "a",
  "A",
  "aa",
  "aA",
  "Aa",
  "AA",
  "ab",
  "aB",
  "Ab",
  "AB",
  "b",
  "B",
  "BB",
  "BBB",
  "c",
  "C",
  "d",
  "D",
  "dd",
  "~",
  "中",
]
`;

exports[`segmentSorter 0 0-21 should sort strings 1`] = `
[
  ".",
  "..",
  "_",
  "A",
  "a",
  "AA",
  "Aa",
  "aA",
  "aa",
  "AB",
  "Ab",
  "aB",
  "ab",
  "B",
  "b",
  "BB",
  "BBB",
  "C",
  "c",
  "D",
  "d",
  "dd",
  "~",
  "中",
]
`;

exports[`segmentSorter 1 1-0 1-0-2 should sort strings 1`] = `
[
  ".",
  "..",
  "a",
  "aa",
  "ab",
  "aA",
  "aB",
  "b",
  "c",
  "d",
  "dd",
  "_",
  "A",
  "Aa",
  "Ab",
  "AA",
  "AB",
  "B",
  "BB",
  "BBB",
  "C",
  "D",
  "~",
  "中",
]
`;

exports[`segmentSorter 1 1-0 should sort strings 1`] = `
[
  ".",
  "..",
  "a",
  "aa",
  "ab",
  "aA",
  "aB",
  "b",
  "c",
  "d",
  "dd",
  "_",
  "A",
  "Aa",
  "Ab",
  "AA",
  "AB",
  "B",
  "BB",
  "BBB",
  "C",
  "D",
  "~",
  "中",
]
`;

exports[`segmentSorter 1 1-2 1-2-0 should sort strings 1`] = `
[
  ".",
  "..",
  "a",
  "aa",
  "ab",
  "aA",
  "aB",
  "b",
  "c",
  "d",
  "dd",
  "A",
  "Aa",
  "Ab",
  "AA",
  "AB",
  "B",
  "BB",
  "BBB",
  "C",
  "D",
  "_",
  "~",
  "中",
]
`;

exports[`segmentSorter 1 1-2 should sort strings 1`] = `
[
  ".",
  "..",
  "a",
  "aa",
  "ab",
  "aA",
  "aB",
  "b",
  "c",
  "d",
  "dd",
  "A",
  "Aa",
  "Ab",
  "AA",
  "AB",
  "B",
  "BB",
  "BBB",
  "C",
  "D",
  "_",
  "~",
  "中",
]
`;

exports[`segmentSorter 2 2-0 2-0-1 should sort strings 1`] = `
[
  ".",
  "..",
  "A",
  "AA",
  "AB",
  "Aa",
  "Ab",
  "B",
  "BB",
  "BBB",
  "C",
  "D",
  "_",
  "a",
  "aA",
  "aB",
  "aa",
  "ab",
  "b",
  "c",
  "d",
  "dd",
  "~",
  "中",
]
`;

exports[`segmentSorter 2 2-0 should sort strings 1`] = `
[
  ".",
  "..",
  "A",
  "AA",
  "AB",
  "Aa",
  "Ab",
  "B",
  "BB",
  "BBB",
  "C",
  "D",
  "_",
  "a",
  "aA",
  "aB",
  "aa",
  "ab",
  "b",
  "c",
  "d",
  "dd",
  "~",
  "中",
]
`;

exports[`segmentSorter 2 2-1 2-1-0 should sort strings 1`] = `
[
  ".",
  "..",
  "A",
  "AA",
  "AB",
  "Aa",
  "Ab",
  "B",
  "BB",
  "BBB",
  "C",
  "D",
  "a",
  "aA",
  "aB",
  "aa",
  "ab",
  "b",
  "c",
  "d",
  "dd",
  "_",
  "~",
  "中",
]
`;

exports[`segmentSorter 2 2-1 should sort strings 1`] = `
[
  ".",
  "..",
  "A",
  "AA",
  "AB",
  "Aa",
  "Ab",
  "B",
  "BB",
  "BBB",
  "C",
  "D",
  "a",
  "aA",
  "aB",
  "aa",
  "ab",
  "b",
  "c",
  "d",
  "dd",
  "_",
  "~",
  "中",
]
`;

exports[`segmentSorter 12 12-0 should sort strings 1`] = `
[
  ".",
  "..",
  "a",
  "A",
  "aa",
  "aA",
  "Aa",
  "AA",
  "ab",
  "aB",
  "Ab",
  "AB",
  "b",
  "B",
  "BB",
  "BBB",
  "c",
  "C",
  "d",
  "D",
  "dd",
  "_",
  "~",
  "中",
]
`;

exports[`segmentSorter 12 should sort strings 1`] = `
[
  ".",
  "..",
  "a",
  "A",
  "aa",
  "aA",
  "Aa",
  "AA",
  "ab",
  "aB",
  "Ab",
  "AB",
  "b",
  "B",
  "BB",
  "BBB",
  "c",
  "C",
  "d",
  "D",
  "dd",
  "_",
  "~",
  "中",
]
`;

exports[`segmentSorter 21 21-0 should sort strings 1`] = `
[
  ".",
  "..",
  "A",
  "a",
  "AA",
  "Aa",
  "aA",
  "aa",
  "AB",
  "Ab",
  "aB",
  "ab",
  "B",
  "b",
  "BB",
  "BBB",
  "C",
  "c",
  "D",
  "d",
  "dd",
  "_",
  "~",
  "中",
]
`;

exports[`segmentSorter 21 should sort strings 1`] = `
[
  ".",
  "..",
  "A",
  "a",
  "AA",
  "Aa",
  "aA",
  "aa",
  "AB",
  "Ab",
  "aB",
  "ab",
  "B",
  "b",
  "BB",
  "BBB",
  "C",
  "c",
  "D",
  "d",
  "dd",
  "_",
  "~",
  "中",
]
`;
